sdk/src/test/scala/com/akkaserverless/javasdk/impl/eventsourcedentity/EventSourcedEntitiesImplSpec.scala

```diff
@@ -287,50 +255,64 @@ object EventSourcedEntitiesImplSpec {
-      val cart = mutable.Map.empty[String, Item]
+    class TestCart(@EntityId val entityId: String) extends EventSourcedEntityBase[ShoppingCartDomain.Cart] {
+
+      override def emptyState(): ShoppingCartDomain.Cart = ShoppingCartDomain.Cart.getDefaultInstance
 
       @CommandHandler
-      def getCart: ShoppingCartApi.Cart = Protocol.cart(cart.values.toSeq: _*)
+      def getCart(currentState: ShoppingCartDomain.Cart,
+                  command: ShoppingCartApi.GetShoppingCart): Effect[ShoppingCartApi.Cart] =
+        effects().reply(Protocol.cart(domainCartToMap(currentState).values.toList: _*))
 
       @CommandHandler
-      def addItem(item: ShoppingCartApi.AddLineItem, ctx: CommandContext): Empty = {
-        if (item.getQuantity == -42) {
-          // emit and then fail on magic negative quantity, for testing atomicity
-          ctx.emit(Protocol.itemAdded(item.getProductId, item.getName, item.getQuantity))
+      def addItem(currentState: ShoppingCartDomain.Cart, command: ShoppingCartApi.AddLineItem): Effect[Empty] = {
+        if (command.getQuantity <= 0) {
+          effects().error(s"Cannot add negative quantity of item [${command.getProductId}]")
+        } else {
+          effects()
+            .emitEvent(Protocol.itemAdded(command.getProductId, command.getName, command.getQuantity))
+            .thenReply(_ => Empty.getDefaultInstance)
         }
-        if (item.getQuantity <= 0) ctx.fail(s"Cannot add negative quantity of item [${item.getProductId}]")
-        ctx.emit(Protocol.itemAdded(item.getProductId, item.getName, item.getQuantity))
-        Empty.getDefaultInstance
       }
 
       @EventHandler
-      def itemAdded(itemAdded: ShoppingCartDomain.ItemAdded): Unit = {
-        if (itemAdded.getItem.getName == "FAIL") throw new RuntimeException("Boom: name is FAIL") // fail for testing
-        val currentQuantity = cart.get(itemAdded.getItem.getProductId).map(_.quantity).getOrElse(0)
-        cart.update(itemAdded.getItem.getProductId,
-                    Item(itemAdded.getItem.getProductId,
-                         itemAdded.getItem.getName,
-                         currentQuantity + itemAdded.getItem.getQuantity))
+      def itemAdded(currentState: ShoppingCartDomain.Cart,
+                    command: ShoppingCartDomain.ItemAdded): ShoppingCartDomain.Cart = {
+        val updatedItem = command.getItem
+        if (updatedItem.getName == "FAIL") throw new RuntimeException("Boom: name is FAIL") // fail for testing
+        val currentItems = domainCartToMap(currentState)
+        val updatedItems = currentItems.updatedWith(
+          updatedItem.getProductId
+        ) {
+          case Some(existing) =>
+            Some(Item(updatedItem.getProductId, updatedItem.getName, existing.quantity + updatedItem.getQuantity))
+          case None => Some(Item(command.getItem.getProductId, command.getItem.getName, command.getItem.getQuantity))
+        }
+        domainCartFromMap(updatedItems)
```

Comment:
`(State, Command) => Effect` and `(State, Event) => State` looks so much better. 

IMO, it makes very clear what is the expected implementation. Especially for the event handler.

`itemAdded(itemAdded: ShoppingCartDomain.ItemAdded): Unit` says nothing. Could be anything.

Reply:
I agree 👍